torch/csrc/dynamo/guards.cpp

```diff
@@ -574,6 +574,367 @@ static struct PyModuleDef _module = {
+    this->failure_reason = failure_reason;
+  }
+
+  std::string repr() {
+    return "GuardDebugInfo(num_guards_executed=" +
+        std::to_string(num_guards_executed) +
+        ", failure_reason=" + failure_reason + ")";
+  }
+
+  // Failure reason for a leaf guard.
+  std::string failure_reason;
+
+  // Total number of executed guards so far.
+  int num_guards_executed;
+};
+
+/**
+ * Base class for the terminal or leaf guard in the GuardManager hierarchy.
+ */
+class LeafGuard {
+ public:
+  // Runs the guard and prepares a GuardDebugInfo object.
+  std::pair<bool, GuardDebugInfo> check_with_debug_info(py::object value) {
+    bool result = run_guards(value);
+    if (result == false) {
+      std::string reason = get_failure_reason(value);
+      return std::make_pair(result, GuardDebugInfo(0, reason));
+    }
+    return std::make_pair(result, GuardDebugInfo(0, "PASS"));
+  }
+
+  // Runs the guard
+  bool check(py::object value) {
+    return run_guards(value);
+  }
+
+  virtual bool run_guards(py::object value) = 0;
+  virtual std::string get_failure_reason(py::object value) = 0;
+  virtual ~LeafGuard() = default;
+};
+
+/**
+ * Represents a leaf guard that accepts the python guard check function. We
+ * would like to have most of the guards in C++ (to avoid a Python function
+ * call).  But, it will take some time to reach that goal. Also, there might be
+ * cases where its too tedious to write an equivalent C++ guard.
+ *
+ * PythonLambdaGuard allows us to gradually move to C++. We can start from all
+ * guards of type PythonLambaGuard and incrementally move expensive guards to
+ * C++.
+ */
+class PythonLambdaGuard : public LeafGuard {
+ public:
+  PythonLambdaGuard(py::object guard_check_fn, py::object print_failure_fn) {
+    if (py::isinstance<py::function>(guard_check_fn) &&
+        py::isinstance<py::function>(print_failure_fn)) {
+      _guard_check_fn = py::cast<py::function>(guard_check_fn);
+      _print_failure_fn = py::cast<py::function>(print_failure_fn);
+    } else {
+      throw py::type_error("PythonLambdaGuard expects callables");
+    }
+  }
+
+  // Runs the lambda function with the current f_locals value.
+  bool run_guards(py::object value) override {
+    return py::cast<bool>(_guard_check_fn(value));
+  }
+
+  std::string get_failure_reason(py::object value) override {
+    return py::cast<std::string>(_print_failure_fn(value));
+  }
+
+ private:
+  // The user provided lambda function for check_fn.
+  py::function _guard_check_fn;
+  // The user provided lambda function to get guard failure reason.
+  py::function _print_failure_fn;
+};
+
+/**
+ * Base class representing how to access the child value from the py::object
+ * given to the parent check function.
+ *
+ * These will have one to one mapping with Source objects in Dynamo. Source
+ * provides a way to access the underlying object using its name() function.
+ * GuardAccessor is basically its C++ tree equivalent.
```

Comment:
Why do we need C++ dynamo sources? I was thinking we don't need sources in C++... since they are just used for guard construction. Once the guard FN is made, you can just pass it down to C++ as a string.